### PR TITLE
[INFINITY-1417] Refactor custom recovery

### PR DESCRIPTION
**Pull request by gabrielhartmann**
The goal is to make it substantially easier to provide custom recovery logic for permanently failed pods.

There are perhaps 3 major portions of this change:
1. Use `PodInstanceRequirement` instead of `String` to determine whether Steps in different plans are conflicting (dirty asset management)
1. Provide the `RecoveryPlanOverrider` interface which allows for the custom overriding of behavior with a pod -> phase mapping of overriding behavior.
1. No longer provide the capability to easily replace the whole RecoveryManager in favor of the mechanism above.

An additional improvement which came out of this work was a clearer handling of pods which have permanently failed.  They are now marked as failed in a consistent fashion.  This mark is cleared when a step which was recovering a permanently failed pod completes.  Please let me know if I can clarify anything.  I tried decomposing this PR but ran into difficulties not worth the time.

I haven't cleaned up all comments etc. yet, but please feel free to nitpick away.

EDIT:  For ease of review the commits aren't totally terrible, so maybe stepping through them would be helpful.

### CODE REVIEW

File: frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsRecoveryPlanManager.java [hunk 1]
```diff
@@ -137,24 +137,29 @@ private void setNameNodePhase(Phase phase) {
     private Phase initNNRecoveryPhase(Plan inputPlan, int index) {
         Phase inputPhase = inputPlan.getChildren().get(0);
         int offset = index * 2;
+
         Step inputBootstrapStep = inputPhase.getChildren().get(offset + 0);
+        PodInstanceRequirement bootstrapPpodInstanceRequirement = PodInstanceRequirement.create(
```
Comment (mohitsoni) [inline frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsRecoveryPlanManager.java]: ultraNit: `s/bootstrapPpodInstanceRequirement/bootstrapPodInstanceRequirement/g`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/recovery/DefaultRecoveryStep.java [hunk 2]
```diff
@@ -6,60 +6,60 @@
...
 
     public RecoveryType getRecoveryType() {
-        return recoveryType;
+        return podInstanceRequirement.getRecoveryType();
     }
 
     @Override
     public String getMessage() {
-        return super.getMessage() + " RecoveryType: " + recoveryType.name();
+        return super.getMessage() + " RecoveryType: " + getRecoveryType().name();
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/recovery/DefaultRecoveryStep.java]: `String.format()` here imo

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/PodInstance.java [hunk 3]
```diff
@@ -28,4 +28,10 @@ static String getName(PodSpec podSpec, int index) {
     public static String getName(String podType, int index) {
         return String.format("%s-%d", podType, index);
     }
+
+    default boolean conflicts(PodInstance podInstance) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/PodInstance.java]: same `conflictsWith` comment as above

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/recovery/monitor/TimedFailureMonitor.java [hunk 4]
```diff
@@ -69,7 +78,12 @@ public boolean hasFailed(TaskInfo terminatedTask) {
                 + taskExpiredTime + " which is " + now.after(taskExpiredTime));
 
         if (now.after(taskExpiredTime)) {
-            FailureUtils.markFailed(terminatedTask);
+            try {
+                PodInstance podInstance = TaskUtils.getPodInstance(configStore, terminatedTask);
+                FailureUtils.markFailed(podInstance, stateStore);
+            } catch (TaskException e) {
+                log.error("Failed to get pod instance", e);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/recovery/monitor/TimedFailureMonitor.java]: - could say "Failed to get pod instance to mark as failed" to give more context in logs
- could log the relevant task in this error (unless it's already logged in the exception...)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PodInstanceRequirement.java [hunk 5]
```diff
@@ -88,11 +127,20 @@ public PodInstance getPodInstance() {
...
-     * should be created, even though (now failed) TaskInfos are still present in the StateStore.
-     */
-    public boolean isPermanentReplacement() {
-        return isPermanentReplacement;
+    public RecoveryType getRecoveryType() {
+        return recoveryType;
+    }
+
+    @Override
+    public String toString() {
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PodInstanceRequirement.java]: Seconded. Pretty uncommon, except for things that can be serialized to and from strings.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/recovery/RecoveryPlanOverrider.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package com.mesosphere.sdk.scheduler.recovery;
+
+import com.mesosphere.sdk.scheduler.plan.Phase;
+import com.mesosphere.sdk.scheduler.plan.PodInstanceRequirement;
+
+import java.util.Optional;
+
+/**
+ * Created by gabriel on 5/1/17.
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/recovery/RecoveryPlanOverrider.java]: Javadoc

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 7]
```diff
@@ -275,10 +275,10 @@ public Builder setPlans(Collection<Plan> plans) {
 
         /**
          * Sets the provided {@link PlanManager} to be the plan manager used for recovery.
-         * @param recoveryManagerFactory the factory whcih generates the custom recovery plan manager
+         * @param recoveryPlanOverriderFactory the factory whcih generates the custom recovery plan manager
          */
-        public Builder setRecoveryManagerFactory(RecoveryPlanManagerFactory recoveryManagerFactory) {
-            this.recoveryPlanManagerFactory = recoveryManagerFactory;
+        public Builder setRecoveryManagerFactory(RecoveryPlanOverriderFactory recoveryPlanOverriderFactory) {
+            this.recoveryPlanOverriderFactor = recoveryPlanOverriderFactory;
```
Comment (mohitsoni) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: `s/recoveryPlanOverriderFactor/recoveryPlanOverriderFactory/g`

**Comment by gabrielhartmann** [comment]
@benclarkwood @nickbp:  All required comments addressed.  Please re-review.

**Comment by gabrielhartmann** [comment]
retest this please

**Comment by gabrielhartmann** [comment]
retest this please

**Pull request merged**